### ISSUE TITLE
Port the code to C++11

### ISSUE THREAD

**Issue by seladb** [seladb/PcapPlusPlus#977]
We recently moved to C++11 and big parts of the code are still using C++98 style.
This will be an ongoing task to port more and more parts of the code to C++11.

For example:
- Use `nullptr` instead of `NULL`
- Use `auto` where possible
- Use `enum class` more
- Use range-based loops where possible
- Use `delete`, `default`, `override` where needed
- Use `unordered_map` instead of `map` for better performance
- etc.

**Comment by InfiniteVerma** [comment]
I'd like to help with this.

**Comment by egecetin** [comment]
Not directly related with C++11 but also, library have some many C style casts, which can be easily located by using `cppcheck`. Anyone who wants fix them just remove or comment `cstyleCast:*` line from `cppcheckSuppressions.txt` and run `pre-commit run --all-files` or manually run `cppcheck` from command line.

**Comment by seladb** [comment]
> Hi @seladb - I would like to collaborate on this task. Is there any task breakdown already or is there any progress? Also, why porting to C++11 instead of later versions? (is it because the latest port you mention in your first comment was to C++11?
> 
> UPDATE - based on the Release notes (changes from v22.05) it seems this issue/ticket should be closed...

Hi @jpcofr , thanks for offering your help! Actually, the reason this issue is still open is because this is an ongoing effort. We migrated to C++11 about a year ago, but a lot of the code base is still in C++98 style (definitely most of the code that was written more than a year ago). For example: we don't always use `auto` where needed, we still use `std::map` where we could use `std::unordered_map`, we don't use range-based loops in many cases, etc. We could use your help with those.

As I mentioned above - we don't expect a huge PR that fixes all of these things. We actually prefer small PRs that make incremental fixes, and we hope that we can slowly migrate as much of the code as possible.

It'd be really great if you can help. I'd like to thank you again and I'm looking forward to seeing your contributions! 🙏 🙏

**Comment by jpcofr** [comment]
>  be really great if you can help. 

@seladb ok I can collaborate. I am new to this repo so I need to check out how is everything working. What would be the next namespace/class etc. that you would like to be ported? I can start from there...

**Comment by seladb** [comment]
> > be really great if you can help.
> 
> @seladb ok I can collaborate. I am new to this repo so I need to check out how is everything working. What would be the next namespace/class etc. that you would like to be ported? I can start from there...

@jpcofr If you want you can start with [`HttpAnalyzer`](https://github.com/seladb/PcapPlusPlus/tree/master/Examples/HttpAnalyzer). It's a quite simple utility (or an example of how to use PcapPlusPlus), and it uses a `std::map` that can be converted to `std::unordered_map`. It also has quite a few for loops that can be modified, and I'm sure we can use the `auto` keyword more.
This utility has [python-based tests](https://github.com/seladb/PcapPlusPlus/blob/master/Tests/ExamplesTest/tests/test_httpanalyzer.py) that you can run locally, and they also run in GitHub Actions as part of the CI: https://github.com/seladb/PcapPlusPlus/actions/runs/5988356929/job/16243457492#step:8:65

Please let me know if you need any help

**Comment by tigercosmos** [comment]
btw, maybe we can change the issue to C++20 directly.

**Comment by seladb** [comment]
> btw, maybe we can change the issue to C++20 directly.

@tigercosmos for now I prefer to keep supporting C++11 to support a broader variety of platforms and devices

**Comment by Dimi1010** [comment]
Another thing is replacing `ostringsream` usages with `to_string` where it is used to convert only a single value to string.

**Comment by rtmeng** [comment]
Hi @seladb , I would also like to collaborate on this, is it still open?

**Comment by seladb** [comment]
> Hi @seladb , I would also like to collaborate on this, is it still open?

Yes sure, you're more than welcome to contribute! 🙏
Feel free to open PRs and we'll review

**Comment by rtmeng** [comment]
> > Hi @seladb , I would also like to collaborate on this, is it still open?
> 
> Yes sure, you're more than welcome to contribute! 🙏 Feel free to open PRs and we'll review

Great! I'm still looking through the repo but is there anywhere I should start for what should be ported next?

**Comment by tigercosmos** [comment]
> > > Hi @seladb , I would also like to collaborate on this, is it still open?
> > 
> > 
> > Yes sure, you're more than welcome to contribute! 🙏 Feel free to open PRs and we'll review
> 
> Great! I'm still looking through the repo but is there anywhere I should start for what should be ported next?

No, you can start at any place that uses C++98  instead of C++11. You can also refer to some merged PR for references.

**Comment by seladb** [comment]
Hi @rtmeng ! I just replied to your email. If you're interested in something nice and simple to get into the project, you can try modernizing the [tutorials](https://github.com/seladb/PcapPlusPlus/tree/master/Examples/Tutorials) code

**Comment by tigercosmos** [comment]
@Dimi1010 What are your thoughts on how much we still have to finish?

**Comment by Dimi1010** [comment]
> @Dimi1010 What are your thoughts on how much we still have to finish?

@tigercosmos 

I think the `map` -> `unordered_map` is almost done. The only place I can find a `#include <map>` is in `TcpReassembly.h` and `TextBasedProtocol.h`

For C-casts, I clean them up as I find them, but there are still plenty to go around... and it sucks you can't find them easily with a find command...

Two things I might add to the list:
- Increased preference on using `unique_ptr` and `shared_ptr` over the raw pointers for dynamic memory management.
- Increased preference of "pass-by-value + `std::move`" instead of `const T&` for constructors that just copy the ctor parameter into member fields, as that would allow moving the parameter into the member field via double move then copying even when its not strictly required. Reasoning: https://stackoverflow.com/a/51706522/11922936
- - Setup of perfect forwarding is also an option, but its more of a hassle and requires templates.

**Comment by egecetin** [comment]
@Dimi1010 For c style casts you can use `cppcheck`. It should point most of them. Just remove `cstyleCast:*` line from `cppcheckSuppressions.txt` and run pre-commit locally or directly cppcheck. It is up to you

**Comment by lumiZGorlic** [comment]
Hi, i am trying to familiarise myself with the code base so i can later contribute with some tasks. This looks like a good way to start so i'd be happy to port some code to C++11 if that's OK ? I'll look at files in Packet++ first. if there's anything specific that should be done first please let me know.

**Comment by seladb** [comment]
> Hi, i am trying to familiarise myself with the code base so i can later contribute with some tasks. This looks like a good way to start so i'd be happy to port some code to C++11 if that's OK ? I'll look at files in Packet++ first. if there's anything specific that should be done first please let me know.

Hi @lumiZGorlic ! Thank you so much for considering contributing to this project! 🙏 
Packet++ is a good place to start. You can choose whatever protocol (layer) you want, I'd start with the simpler ones to familiarize yourself with the code-base. Please feel free to reach out to use if you have any question

**Comment by tigercosmos** [comment]
@lumiZGorlic The code under `Examples` also has room for improvement, and it should be easier for beginners. For example, you can start with refactoring the C-style casting to C++ casting first.

**Comment by DeepakReddy1999** [comment]
Hi @seladb ,I would like to contribute for the Examples and get familiar with the code base

**Comment by seladb** [comment]
> Hi @seladb ,I would like to contribute for the Examples and get familiar with the code base 

Sure @DeepakReddy1999 you're more than welcome to contribute! 🙏

**Comment by DeepakReddy1999** [comment]
Hi @seladb .There are still some NULLs in the code. Should we change it?

**Comment by tigercosmos** [comment]
@DeepakReddy1999  Yes, please go ahead.

**Comment by seladb** [comment]
> Hi @seladb .There are still some NULLs in the code. Should we change it?

There are very few instances of `NULL` left, mostly in `DpdkTests` and `KniTests`. Please notice that we decided to keep `NULL` when calling Windows API like here: https://github.com/seladb/PcapPlusPlus/blob/master/Examples/IcmpFileTransfer/IcmpFileTransfer-pitcher.cpp#L40-L45

**Comment by egecetin** [comment]
For Windows API related NULLs, maybe a comment should be added to all of them to prevent future changes? With reference to MSDN page @seladb shared recently.

**Comment by tigercosmos** [comment]
I agree with @egecetin.

**Comment by Dimi1010** [comment]
@seladb What are your thoughts on changing the callbacks from raw function pointers to `std::function`?
It would generally allow more C++11 constructs to be used as callbacks (lambdas, functor objects, etc), although it means 1 more pointer dereference per call and slightly heavier construction on assignment due to type erasure.

cc: @tigercosmos @egecetin @clementperon

**Comment by tigercosmos** [comment]
`std::fucntion` is generally heavy. Maybe we shouldn't do this.

**Comment by egecetin** [comment]
Or maybe just keep high performance parts (DPDK PF_RING etc) with raw and replace others 🤔

**Comment by tigercosmos** [comment]
I guess we can consider the whole PCPP project to be high-performance?

**Comment by Dimi1010** [comment]
`PcapLiveDevice` already uses them for callbacks. I agree with the notion that it may be better to keep them out of the more performance focused library wrappers.

**Comment by gvanem** [comment]
Another lame C++ error:
```c
3rdParty\Getopt-for-Visual-Studio\getopt.h(126): error C2440: 'initializing': cannot convert from 
'const char [1]' to 'char *'
```
easily fixed with:
```diff
--- a/3rdParty/Getopt-for-Visual-Studio/getopt.h 2025-02-15 11:10:13
+++ b/3rdParty/Getopt-for-Visual-Studio/getopt.h 2025-02-15 17:20:34
@@ -110,11 +110,7 @@
 extern char __declspec(dllimport) *__progname;
 #endif

-#ifdef __CYGWIN__
 static char EMSG[] = "";
-#else
-#define        EMSG            ""
-#endif
```

Tested with `cl -std:c++20`. No other issues.

**Comment by tigercosmos** [comment]
@seladb @Dimi1010  @egecetin  @clementperon  As it is 2025 now, maybe consider closing this issue, and start to support c++17 in the next release?

**Comment by seladb** [comment]
> [@seladb](https://github.com/seladb) [@Dimi1010](https://github.com/Dimi1010) [@egecetin](https://github.com/egecetin) [@clementperon](https://github.com/clementperon) As it is 2025 now, maybe consider closing this issue, and start to support c++17 in the next release?

@tigercosmos Maybe we can rename it to `Port the code to C++14`? We officially moved to C++14 in this PR: https://github.com/seladb/PcapPlusPlus/pull/1821

**Comment by tigercosmos** [comment]
@seladb  I see. I forgot that PR. However, what stops us from moving on to C++17 directly?

**Comment by seladb** [comment]
> [@seladb](https://github.com/seladb) I see. I forgot that PR. However, what stops us from moving on to C++17 directly?

We just moved to C++14 and haven't published a release with it yet. I suggest we give it some time before moving to C++17

**Comment by tigercosmos** [comment]
@seladb  Sure. I will open a new issue for C++14.
